RFC-0011-InferenceMode.md

```diff
@@ -0,0 +1,138 @@
+       "consider making a clone first. ",
+       "If you have a valid use case, please make a feature request to PyTorch.");
+     {
+       at::AutoDispatchBelowInplaceOrView guard;
+       at::redispatch::add_(ks & c10::after_InplaceOrView_keyset, self, other, alpha);
+     }
+     increment_version(self);
+     return self;
+   }
+
+
+   Tensor expand(c10::DispatchKeySet ks, const Tensor & self, IntArrayRef size, bool implicit) {
+
+     TORCH_CHECK(c10::impl::is_all_dispatch_keyset_excluded(c10::autograd_dispatch_keyset),
+       "Calling inplace/view ops on inference tensor outside InferenceMode is not allowed, ",
+       "consider making a clone first. ",
+       "If you have a valid use case, please make a feature request to PyTorch.");
+     auto _tmp = ([&]() {
+       at::AutoDispatchBelowInplaceOrView guard;
+       return at::redispatch::expand(ks & c10::after_InplaceOrView_keyset, self, size, implicit);
+     })();
+     std::function<at::Tensor(const at::Tensor&)> func=nullptr;
+     if (false || !self.unsafeGetTensorImpl()->support_as_strided()) {
+       auto size_vec = size.vec();
+       func = [=](const at::Tensor& input_base) {
+         return input_base.expand(size_vec, implicit);
+       };
+     }
+     auto result = as_view(/* base */ self, /* output */ _tmp, /* is_bw_differentiable */ true, /* is_fw_differentiable */ true, /* view_func */ func, /* creatio
+     return result;
+   }
+ ```
+ - **Inference mode** can be turned on when you are sure you don't need any autograd computation. This saves the cost of creating autograd graph and `as_view` / `version_counter` setup compared to the normal mode.
+ - **Inference tensor** is defined as a tensor without Autograd **and** InplaceOrView keys on it.
+ - **Normal tensor** has both Autograd & InplaceOrView keys. This includes both `requires_grad=true` and `requires_grad=false` tensors. (see [Ideal end state] section for more details).
+ - Additional notes:
+   - All Inference tensors are created in inference mode, but not all of the tensors created in inference mode are inference tensors. For example, a view of normal tensor created in inference mode is still a normal tensor (but with special `creation_meta`!).
+   - (Autograd & !InplaceOrView) and (!Autogad & InplaceOrView) are invalid states, we don't have such tensors.
+
+# Expected Behavior
+## Implementation:
+1. Inference Mode: InplaceOrView not in included, Autograd in excluded
+2. Normal Mode: InplaceOrView in included, Autograd not in excluded
+3. In VariableType kernel, throw an error if input is inference tensor.
+4. In InplaceOrView kernel, throw an error if Autograd keyset is not in excluded set already.
+5. In VariableType kernel, throw an error if input is a view with `NO_VARIABLE_TYPE_VIEW` creation_meta.
+## Behavior
+| Mode          | Input                                    | Op         | Go through Kernels                        | Produced Output                                            |   |   |
```

Comment:
Is the intention that I can generate all of the entries of the table from the Implementation above?